src/text.c

```diff
@@ -0,0 +1,441 @@
+	size_t shift; // amount to shift chars to the left (from unescaping)
+	SdbList/*<size_t>*/ *path;
+	LoadState state;
+	bool unescape; // whether the prev char was a backslash, i.e. the current one is escaped
+} LoadCtx;
+
+
+// load more data from ctx->f into ctx->buf starting at ctx->pos
+// and update ctx->buf_filled and ctx->eof.
+// may realloc and change ctx->buf and ctx->bufsz.
+static bool load_read_more(LoadCtx *ctx) {
+	// realloc if no space
+	if (ctx->pos >= ctx->bufsz - 1) {
+		size_t newsz = ctx->bufsz + ctx->bufsz;
+		if (newsz < ctx->bufsz || newsz > (size_t)INT_MAX) {
+			return false;
+		}
+		char *newbuf = realloc (ctx->buf, newsz);
+		if (!newbuf) {
+			return false;
+		}
+		ctx->buf = newbuf;
+		ctx->bufsz = newsz;
+	}
+	// read full buffer (leave 1 for null-terminating)
+	int red = read (ctx->fd, ctx->buf + ctx->pos, ctx->bufsz - 1 - ctx->pos);
+	if (red < 0) {
+		// error
+		return false;
+	}
+	// process buffer
+	ctx->buf_filled = ctx->pos + red;
+	ctx->eof = red == 0;
+	return true;
+}
+
+// to be called at the end of a line.
+// save all the data processed from the line into the database.
+static void load_flush_line(LoadCtx *ctx) {
+	ctx->unescape = false;
+	// finish up the line
+	ctx->buf[ctx->pos - ctx->shift] = '\0';
+	switch (ctx->state) {
+	case STATE_PATH: {
+		ls_push (ctx->path, (void *)ctx->token_begin);
+		SdbListIter *it;
+		void *token_off_tmp;
+		ctx->cur_db = ctx->root_db;
+		ls_foreach (ctx->path, it, token_off_tmp) {
+			size_t token_off = (size_t)token_off_tmp;
+			if (!ctx->buf[token_off]) {
+				continue;
+			}
+			ctx->cur_db = sdb_ns (ctx->cur_db, ctx->buf + token_off, 1);
+			if (!ctx->cur_db) {
+				ctx->cur_db = ctx->root_db;
+				break;
+			}
+		}
+		ls_destroy (ctx->path);
+		break;
+	}
+	case STATE_VALUE:
+		if (!*ctx->buf || !ctx->buf[ctx->token_begin]) {
+			break;
+		}
+		sdb_set (ctx->cur_db, ctx->buf, ctx->buf + ctx->token_begin, 0);
+		break;
+	default:
+		break;
+	}
+	if (ctx->buf_filled && ctx->pos < ctx->buf_filled - 1) {
+		memmove (ctx->buf, ctx->buf + ctx->pos + 1, ctx->buf_filled - ctx->pos - 1);
+	}
+	// prepare for next line
+	ctx->shift = 0;
+	if (ctx->pos + 1 <= ctx->buf_filled) {
+		ctx->buf_filled -= ctx->pos + 1;
+	} else {
+		// special case, when at the end of the file
+		// here pos == buf_filled
+		ctx->buf_filled = 0;
+	}
+	ctx->pos = 0;
+	ctx->state = STATE_NEWLINE;
+}
+
+static void load_process_single_char(LoadCtx *ctx) {
+	if (ctx->buf[ctx->pos] == '\n' || ctx->buf[ctx->pos] == '\r') {
+		load_flush_line (ctx);
+		return;
+	}
+
+	if (ctx->state == STATE_NEWLINE) {
+		// at the start of a line, decide whether it's a path or a k=v
+		// by whether there is a leading slash.
+		if (ctx->buf[ctx->pos] == '/') {
+			ctx->state = STATE_PATH;
+			ctx->token_begin = 1;
+			ctx->pos++;
+			return;
+		} else {
+			ctx->state = STATE_KEY;
+		}
+	}
+
+	if (ctx->unescape) {
+		char raw_char;
+		switch (ctx->buf[ctx->pos]) {
```

Comment:
cache ctx->buf[ctx->pos] in char curchar, this is being checked several times below and you dont need to dereference 2 pointers on every if